ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp

```diff
@@ -848,8 +954,372 @@ void adjust_conv_op_config_for_auto_shard_if_necessary(
+        in_channels_padded = tt::round_up(in_channels, conv_config.input_channels_alignment);
+    }
+
+    uint32_t nhw_out_padded_ntile = get_num_cores_nhw_from_parallel_config(output_parallel_config) *
+                                    conv_out_memory_config.shard_spec.value().shape[0] / tt::constants::TILE_HEIGHT;
+
+    OptimizedConvBlockConfig opt_conv_op_block_config = determine_per_core_conv_block_config(
+        input_parallel_config,
+        opt_conv_op_parallel_config,
+        in_channels_padded,
+        nhw_out_padded_ntile,
+        conv_config.act_block_h_override,
+        conv_config.act_block_w_div,
+        kernel_size[0],
+        kernel_size[1],
+        get_fp32_dest_acc_en(compute_config),
+        conv_config.enable_split_reader);
+    return {opt_conv_op_parallel_config, opt_conv_op_block_config, conv_out_memory_config};
+}
+
+std::pair<uint32_t,uint32_t> conv2d::estimate_L1_usage(
+    tt::ARCH arch, TensorMemoryLayout shard_layout,
+    const DataType input_dtype, const DataType weights_dtype, const DataType output_dtype,
+    const DeviceComputeKernelConfig& compute_kernel_config,
+    const OptimizedConvBlockConfig& block_config, const OptimizedConvParallelizationConfig& pconfig,
+    const Shape& input_shape, const Shape& weights_shape, const Shape& output_shape,
+    uint32_t output_channels, uint32_t groups, std::array<uint32_t, 2> kernel_size,
+    const Conv2dConfig& conv_config, const MemoryConfig& output_memory_config, const  bool enable_bias, bool use_non_tile_height
+    ) {
+
+    bool untilize_out = conv_config.output_layout == Layout::ROW_MAJOR;
+    uint32_t batch_size = output_shape[0];
+    uint32_t conv_output_h = output_shape[1];
+    uint32_t conv_output_w = output_shape[2];
+
+    uint32_t input_channels = input_shape[3];
+    bool is_depthwise_conv = groups == input_channels && groups == output_channels;
+
+    uint32_t input_tile_size = tt::tile_size(datatype_to_dataformat_converter(input_dtype));
+    uint32_t weights_tile_size = tt::tile_size(datatype_to_dataformat_converter(weights_dtype));
+    uint32_t bias_tile_size = 0;
+    if(enable_bias) {
+        bias_tile_size = tt::tile_size(datatype_to_dataformat_converter(weights_dtype));
+    }
+    uint32_t output_tile_size = tt::tile_size(datatype_to_dataformat_converter(output_dtype));
+
+    auto [math_fidelity, math_approx_mode, fp32_dest_acc_en, packer_l1_acc, dst_full_sync_en] =
+        get_compute_kernel_config_args(arch, compute_kernel_config);
+
+    uint32_t act_block_w_ntiles = block_config.act_block_w_ntiles;
+    uint32_t act_block_h_ntiles = block_config.act_block_h_ntiles;
+    uint32_t act_block_num_tiles = block_config.act_block_h_ntiles * act_block_w_ntiles;
+
+
+    uint32_t weight_matrix_height = weights_shape.padded_shape()[2];
+    uint32_t weight_matrix_width = weights_shape.padded_shape()[3];
+    uint32_t weight_matrix_height_ntiles = weight_matrix_height / tt::constants::TILE_HEIGHT;
```

Comment:
For all divisions in this function, do we take care of making sure its not fractional, or need to take ceiling?

Reply:
For these two calculations, we take the padded shape, so it should be divisible by tile size. 

Other places where division is used, those values are expected to be divisible. For example, weight matrix width and per core out width.